### PR TITLE
Account for the window borders when restoring from fullscreen

### LINKED ISSUE #10199: Quake Mode has incorrect window size after toggling fullscreen mode
### Windows Terminal version (or Windows build number)

Windows: 10.0.19043.0, Terminal: 1.9.1445.0

### Other Software

I am using 125 % DPI scaling

### Steps to reproduce

1. Invoke Quake Mode overlay
2. Press F11
3. Press F11 again

https://user-images.githubusercontent.com/1827784/119619525-61822d80-be04-11eb-91c1-004cdaefc9b1.mp4

### Expected Behavior

After leaving fullscreen mode, the quake bar should have the same size as before

### Actual Behavior

The window is moved to the right, there is a very noticeable gap between the window and the border of the screen
image

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

When we're restoring from fullscreen, we do a little adjustment to make sure to clamp the window bounds within the bounds of the active monitor. We unfortunately didn't account for the size of the non-client area (the invisible borders around our 1px border). This didn't matter most of the time, but if the window was within ~8px of the side of the monitor (any side), then restoring from fullscreen would actually move it to the wrong place. 

As it turns out, the `_quake` window is within ~8px of the edges of the monitor _very often_.

## References
* regressed in #9737

## PR Checklist
* [x] Closes #10199
* [x] I work here
* [ ] Tests added/passed
* [n/a] Requires documentation to be updated

## Validation Steps Performed
The repro in the bug was fairly straightforward. It doesn't happen anymore.

### CODE REVIEW

**Review by DHowett — APPROVED** [review]

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 1]
```diff
@@ -884,23 +884,36 @@ void IslandWindow::_RestoreFullscreenPosition(const RECT rcWork)
...
 
+    const til::size ncSize{ GetTotalNonClientExclusiveSize(dpiWindow) };
+
+    RECT rcWorkAdjusted = rcWork;
+
+    // GH#10199 - adjust the size of the "work" rect by the size of our borders.
+    // We want to make sure the window is restored within the bounds of the
+    // monitor we're on, but it's totally fine if the invisible borders are
+    // outside the monitor.
+    rcWorkAdjusted.left -= ncSize.width<long>() / 2;
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:895]: The optimizer will probably catch it... but I'm not a huge fan of doing the division twice...

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 2]
```diff
@@ -884,23 +884,36 @@ void IslandWindow::_RestoreFullscreenPosition(const RECT rcWork)
...
+    rcWorkAdjusted.right += ncSize.width<long>() / 2;
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:896]: Could width or height reasonably be an odd number where this won't work properly? Or will the offsetting below correct that to slide the whole thing 1px into place at the end then?

I think this is fine... but this is where my brain is tying itself into a knot.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:895]: meh, okay

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:896]: Form my experience, these numbers are always (8 * dpiScale), so I don't think they can be. IIRC the borders from DWM (the ones that are causing the problem here) are always "8px"

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 3]
```diff
@@ -884,23 +884,36 @@ void IslandWindow::_RestoreFullscreenPosition(const RECT rcWork)
...
+    if (rcRestore.right > rcWorkAdjusted.right)
     {
-        OffsetRect(&rcRestore, rcWork.right - rcRestore.right, 0);
+        OffsetRect(&rcRestore, rcWorkAdjusted.right - rcRestore.right, 0);
     }
-    if (rcRestore.left < rcWork.left)
+    if (rcRestore.left < rcWorkAdjusted.left)
     {
-        OffsetRect(&rcRestore, rcWork.left - rcRestore.left, 0);
+        OffsetRect(&rcRestore, rcWorkAdjusted.left - rcRestore.left, 0);
```
Comment (lhecker) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:908]: This fails to work properly if `rcRestore` covers a larger area than `rcWorkAdjusted` doesn't it?

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 2 above)
Comment (miniksa) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:896]: Oh well pffsh. OK.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:895]: I mean whatever. I'm not your mom.

**Review by miniksa — APPROVED** [review]

**Comment by miniksa** [comment]
No automerge so Mike can decide whether I am or am not his mom and also respond to Leonard's comment.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 3 above)
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:908]: I don't think so... I think the way it already was, the whole point was clipping `rcRestore` inside of `rcWork`. So I think it only _ever_ did something when `rcWork` was smaller than `rcRestore`.

Now we're just making `rcWork` a little bigger, so we can keep the invisible borders off the monitor

**Pull request merged**